src/util/bip152.rs

```diff
@@ -0,0 +1,433 @@
+        }
+
+        Ok(HeaderAndShortIds {
+            header: block.header.clone(),
+            nonce: nonce,
+            // Provide coinbase prefilled.
+            prefilled_txs: prefilled,
+            short_ids: short_ids,
+        })
+    }
+}
+
+/// A [BlockTransactionsRequest] structure is used to list transaction indexes
+/// in a block being requested.
+#[derive(PartialEq, Eq, Clone, Debug, PartialOrd, Ord, Hash)]
+pub struct BlockTransactionsRequest {
+    ///  The blockhash of the block which the transactions being requested are in.
+    pub block_hash: BlockHash,
+    ///  The indexes of the transactions being requested in the block.
+    pub indexes: Vec<u64>,
+}
+
+impl Encodable for BlockTransactionsRequest {
+    fn consensus_encode<S: io::Write>(&self, mut s: S) -> Result<usize, io::Error> {
+        let mut len = self.block_hash.consensus_encode(&mut s)?;
+        // Manually encode indexes because they are differentially encoded VarInts.
+        len += VarInt(self.indexes.len() as u64).consensus_encode(&mut s)?;
+        let mut last_idx = 0;
+        for idx in &self.indexes {
+            len += VarInt(*idx - last_idx).consensus_encode(&mut s)?;
+            last_idx = *idx + 1;
+        }
+        Ok(len)
+    }
+}
+
+impl Decodable for BlockTransactionsRequest {
+    fn consensus_decode<D: io::Read>(mut d: D) -> Result<BlockTransactionsRequest, encode::Error> {
+        Ok(BlockTransactionsRequest {
+            block_hash: BlockHash::consensus_decode(&mut d)?,
+            indexes: {
+                // Manually decode indexes because they are differentially encoded VarInts.
+                let nb_indexes = VarInt::consensus_decode(&mut d)?.0 as usize;
+
+                // Since the number of indices ultimately represent transactions,
+                // we can limit the number of indices to the maximum number of
+                // transactions that would be allowed in a vector.
+                let byte_size = (nb_indexes as usize)
+                                    .checked_mul(mem::size_of::<Transaction>())
+                                    .ok_or(encode::Error::ParseFailed("Invalid length"))?;
+                if byte_size > encode::MAX_VEC_SIZE {
+                    return Err(encode::Error::OversizedVectorAllocation {
+                        requested: byte_size,
+                        max: encode::MAX_VEC_SIZE,
+                    });
+                }
+
+                let mut indexes = Vec::with_capacity(nb_indexes);
+                let mut last_index: u64 = 0;
+                for _ in 0..nb_indexes {
+                    let differential: VarInt = Decodable::consensus_decode(&mut d)?;
+                    last_index  = match last_index.checked_add(differential.0 + 1) {
+                        Some(r) => r,
+                        None => return Err(encode::Error::ParseFailed("block index overflow")),
+                    };
+                    indexes.push(last_index);
```

Comment:
for reference: While adding `deserialize(serialize(BlockTransactionsRequest{..}))` test I noticed a subtle off-by-one error here. The increment in L321 (`differential.0 + 1`) is too early. This should only be incremented after we `indexes.push(last_index);`. Otherwise a differentially encoded `[0, 1, 0]` would become `[1, 3, 4]` when it should be `[0, 2, 3]`.

Fixed this in [TODO:].